File: day1/src/graphql/resolvers/bookResolver.ts
import { authors, books } from "../../data/db.js";
import { IResolvers } from "@graphql-tools/utils";

interface Author {
  id: string;
  name: string;
}

interface Book {
  id: string;
  title: string;
  genre: string;
  authorId: string;
}

const bookResolver: IResolvers = {
  Query: {
    getAllBooks: () => books,
    getBook: (_parent, { id }: { id: string }) => books.find((book: Book) => book.id === id),
  },
  Mutation: {
    addBook: (
      _parent,
      { title, genre, authorId }: { title: string; genre: string; authorId: string }
    ) => {
      const newBook: Book = { id: String(books.length + 1), title, genre, authorId };
      books.push(newBook);
      return newBook;
    },
  },
  Book: {
    author: (book: Book) => authors.find((author: Author) => author.id === book.authorId),
  },
};

export default bookResolver;

File: training-module-4-authentication/book-service/src/main/java/com/graphql/book/resolver/BookResolver.java
package com.graphql.book.resolver;

import com.graphql.book.directives.Auth;
import com.graphql.book.model.Book;
import com.netflix.graphql.dgs.*;

import java.util.*;

@DgsComponent
public class BookResolver {

    private static final List<Book> BOOKS = new ArrayList<>(List.of(
            new Book(1L, "GraphQL in Action", 2024, 1L),
            new Book(2L, "Spring Boot with DGS", 2025, 2L)
    ));

    @DgsQuery
    @Auth(requires = "USER")
    public List<Book> getAllBooks() {
        return BOOKS;
    }

    @DgsQuery
    @Auth(requires = "USER")
    public Book getBookById(@InputArgument Long id) {
        return BOOKS.stream().filter(b -> b.id().equals(id)).findFirst().orElse(null);
    }

    @DgsMutation
    @Auth(requires = "ADMIN")
    public Book addBook(@InputArgument String title, @InputArgument int publishedYear, @InputArgument Long authorId) {
        Book newBook = new Book((long) (BOOKS.size() + 1), title, publishedYear, authorId);
        BOOKS.add(newBook);
        return newBook;
    }

    @DgsMutation
    @Auth(requires = "ADMIN")
    public Boolean deleteBook(@InputArgument Long id) {
        return BOOKS.removeIf(b -> b.id().equals(id));
    }
}
